### DISCUSSION TITLE
incoming request

### DISCUSSION THREAD

**Discussion by Elappnano** [miguelgrinberg/Flask-SocketIO#1606] [Q&A]
**Hello
I am using the following code to process incoming request. Although I test the server and client with the random number but when I add get.request the host shows nothing**

**server code:**
```
from flask_socketio import SocketIO, emit 
from flask_socketio import SocketIO, emit 
from flask import Flask, render_template, url_for, copy_current_request_context, request
from random import random
from time import sleep
from threading import Thread, Event


app = Flask(__name__)
app.config['DEBUG'] = True


#turn the flask app into a socketio app
socketio = SocketIO(app, async_mode=None, logger=True, engineio_logger=True)

#random number Generator Thread
thread = Thread()
thread_stop_event = Event()


def randomNumberGenerator():
    # """
    # Generate a random number every 1 second and emit to a socketio instance (broadcast)
    # Ideally to be run in a separate thread?
    # """
    #infinite loop of magical random numbers
    print("Making random numbers")
    while not thread_stop_event.isSet():
        # number = round(random()*10, 3)
		number = request.values.get('test')
        print(number)
        socketio.emit('newnumber', {'number': number}, namespace='/test')
        socketio.sleep(5)

@app.route('/')
def index():
    #only by sending this page first will the client be connected to the socketio instance
    return render_template('index.html')

@socketio.on('connect', namespace='/test')
def test_connect():
    # need visibility of the global thread object
    global thread
    print('Client connected')
    #Start the random number generator thread only if the thread has not been started before.
    if not thread.isAlive():
        print("Starting Thread")
        thread = socketio.start_background_task(randomNumberGenerator)

@socketio.on('disconnect', namespace='/test')
def test_disconnect():
    print('Client disconnected')

if __name__ == '__main__':
    socketio.run(app)
```

**client code:**
```
$(document).ready(function(){
    //connect to the socket server.
    var socket = io.connect('http://' + document.domain + ':' + location.port + '/test');
    var numbers_received = [];

    //receive details from server
    socket.on('newnumber', function(msg) {
        console.log("Received number" + msg.number);
        //maintain a list of ten numbers
        // if (numbers_received.length >= 10){
            // numbers_received.shift()
        // }            
        numbers_received.push(msg.number);
        numbers_string = '';
        for (var i = 0; i < numbers_received.length; i++){
            numbers_string = numbers_string + '<p>' + numbers_received[i].toString() + '</p>';
        }
        $('#log').html(numbers_string);
    });

});
```

I am new to coding in socket.io so I do not know how to process get request.

**Comment by miguelgrinberg** [comment]
What do you mean by this:

> but when I add get.request the host shows nothing

**Comment by Elappnano** [comment]
Hello miguel
I am sending an XMLHttpRequest() to my localhost and want to process it with socket.io when ever the request is sent.
But through the code nothing is being shown on my localhost page.

**Comment by miguelgrinberg** [comment]
Socket.IO does not use XMLHttpRequest. You have to use a Socket.IO client to emit events at the server.

**Comment by Elappnano** [comment]
so it is not possible to get my request with this line in my server code:
`number = request.values.get('test')`
and could you give me an example of how to do it with socket.io client

**Comment by Elappnano** [comment]
As a matter of fact I am getting an XMLHttpRequest to my localhost from another server and want to update my web page without reloading the page.

**Comment by miguelgrinberg** [comment]
You seem to be confusing Socket.IO with standard HTTP requests. If you want your client to communicate with the server using HTTP requests, then you do not need Socket.IO. If you want to use Socket.IO however, then you have to use the `emit()` function in the Socket.IO client or server to communicate with the other side.

**Comment by Elappnano** [comment]
Currently I am using Flask with AJAx as below code to get incoming request. but as you see in my script code I must load my script every one second to check for incoming data and shows it on my page dynamically. I want to run the Ajax script whenever the data available to the serveer and not to check the ajax code every one second. So i am wondering is it able to rewrite my code with socket.io to fulfill this goal.
**Flask code**
```
from flask import Flask, render_template, jsonify, request
import numpy as np
app = Flask(__name__)

values=[]
random_decimal=[]

@app.route('/update_decimal', methods=['POST'])
def updatedecimal():
    value = request.values.get('test')
    values.append(value)
    random_decimal= str(values)
    return jsonify('', render_template('random_decimal_model.html', x=random_decimal))

@app.route('/')
def index():
    value = request.values.get('test')
    values.append(value)
    random_decimal= str(values)
    return render_template('index.html',x=random_decimal)

if __name__ == '__main__':
    app.run(debug=True, port=5000)
```
**Ajax script:**
```
<script src="https://ajax.googleapis.com/ajax/libs/jquery/2.2.0/jquery.min.js"></script> 

<script>
$(function(){
    window.setInterval(function(){
        loadNewDecimal()
    }, 1000)
	
    function loadNewDecimal(){
        $.ajax({
            url:"/update_decimal",
            type: "POST",
            dataType: "json",
            success:function(data){
                $(random_decimal).replaceWith(data)

            }

        });
    }
});
</script>
<div id="random_decimal">
    {{x}}
</div>
```

**Comment by miguelgrinberg** [comment]
The example code in this repository shows how to send data from the server to the client periodically. A background thread is used for that purpose.

**Comment by Elappnano** [comment]
Yeah. the background thread check every one second whether there is data or not and then send it to my client. So in order to use this concept with socket.io how should I rewrite my code?

**Comment by miguelgrinberg** [comment]
@Elappnano I don't understand your question, sorry. If you already have a background thread, you have to add the emit call in the proper place, as the example application in this repository does.
